src/coq_elpi_builtins.ml

```diff
@@ -1209,9 +1212,23 @@ It's a fatal error if Name cannot be located.|})),
     In(gref, "GR",
     COut(closed_ground_term, "Ty",
     Full(global, "reads the type Ty of a global reference."))),
-  (fun gr _ ~depth _ _ state ->
-    let state, ty = type_of_global state gr in
-    state, !:ty, [])),
+  (fun gr _ ~depth { options } _ state ->
+    let state, ty, extra_goals =
+      match options.uinstance with
+      | NoInstance ->
+        let state, (ty, _) = type_of_global state gr None in
+        state, ty, []
+      | ConcreteInstance i ->
+        let state, (ty, _) = type_of_global state gr (Some i) in
+        state, ty, []
+      | VarInstance (v_head, v_args, v_depth) ->
+        let state, (ty, uinst) = type_of_global state gr None in
+        let v' =
+          E.mkUnifVar v_head ~args:(List.map (U.move ~from:v_depth ~to_:depth) v_args) state in
```

Comment:
I think mislead you on zulip. You should move the entire `v` (internally, I think the code will do a no-op on the head and map the arguments as you do, but that is an implementation detail/optim that does not belong here).